### PR TITLE
GH-37655: [C++] Allow joins of large tables in Acero

### LINKED ISSUE #37655: [C++] Acero cannot join large tables because of a misspecified test
**TL;DR: [PR 35087](https://github.com/apache/arrow/pull/35087) introduced an explicit fail in large joins with Acero when key data is larger than 4GB (solving the problem reported by [issue 34474](https://github.com/apache/arrow/issues/34474)). However, I think (but I'm not sure) that this quick fix is too restrictive because the total size condition is applied to the total size of tables to be joined, rather than to the size of keys. As a consequence, Acero fails when trying to merge large tables, even when the size of key data is well below 4 GB.**

-------------------------

__EDIT: It looks like this bug is actually related to the size of the right table, but unrelated to the size of the left table ([see here](https://github.com/apache/arrow/issues/37655#issuecomment-1764857405)).__

In this issue I proceed in four steps. First, I show the test introduced by [PR 35087](https://github.com/apache/arrow/pull/35087) erroneously applies to the total size of data processed in the join, rather than the size of the key data. Second, I try to find the root cause of this behavior. Third, I discuss whether this is really a bug, or the expected behavior of Acero. Fourth, I suggest the sketch of a solution.

**Description of the problem**

[PR 35087](https://github.com/apache/arrow/pull/35087) introduced an explicit fail in large joins with Acero when key data is larger than 4GB ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L476)). However, I discovered that this error message does disappear when I reduce the number of columns in the Arrow Tables I'm trying to merge.

In the following reproducible example, I generate a large Arrow Table and merge it with itself, increasing the number of columns in the right table. The merge works fine for a limited number of columns, then the error message pops up when the number of columns reaches a threshold (8 in my case).

As a consequence, Acero throws an error whenever I try to merge large Arrow Tables, even for tables with key data significantly smaller than 4 GB.

__Reproducible example in `R`__

```
library(stringi)
library(arrow)
library(dplyr)


# Generate a large number of rows with ONE heavy key column
n_rows <- 2e7
length_id <- 20
ids <- stringi::stri_rand_strings(n_rows, length = length_id)

# Create a large Arrow Table with heavy payloads
data <- data.frame(
  id = ids
) |> 
  as_arrow_table() |>
  mutate(
    # Create payload variables
    variable1   = id,
    variable2   = id,
    variable3   = id,
    variable4   = id,
    variable5   = id,
    variable6   = id,
    variable7   = id,
    variable8   = id,
    variable9   = id,
    variable10  = id
  ) |>
  compute()


vars <- names(data)[!names(data) %in% c("id")]
nb_var <- length(vars)

# Join the dataset with itself, with an increasing number of variables
lapply(1:nb_var, function(n) {
  print(paste0("Doing the join with ", n+1, " variables"))
  vars_temp <- c("id", vars[1:n])
  print(vars_temp)
  data_out <- data |> 
    left_join(
      data |>
        select(all_of(vars_temp)),
      by = c("id" = "id")
    ) |>
    compute()
  return("Success!")
})
```

**Cause of the problem**

I dived in the C++ source code of Acero to understand the problem. Disclaimer: I do not know anything about C++, so my report might be messy from here on.

[PR 35087](https://github.com/apache/arrow/pull/35087) introduced a logical test in Status `RowArrayMerge::PrepareForMerge()`. This test computes what I understand to be the size of the sources ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L469C17-L469C17)). 

I think the problem comes from the fact that the Status `RowArrayMerge::PrepareForMerge()` is called twice in `SwissTableForJoinBuild::PreparePrtnMerge()`: once for the keys ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L1333-L1334)) and once for payload variables ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L1355-L1357)). My intuition is that, when applied to payload variables, the logical test actually computes the __size of the payload variables__, so more or less the size of the tables to be joined. __EDIT: It looks like this bug is actually related to the size of the right table, but unrelated to the size of the left table ([see here](https://github.com/apache/arrow/issues/37655#issuecomment-1764857405)).__

**Is this really a bug?**

Given that I don't know how Acero performs joins, I'm not entirely sure whether the 4 GB size limit mentioned by @westonpace in [this message](https://github.com/apache/arrow/issues/34474#issuecomment-1506009490) applies to the size of the keys or to the size of the tables to be joined. My understanding of the discussion in the issue and of the error message is that the size limit applies to keys, so the behavior I describe should be considered as a bug. But maybe I misunderstood, and the size limit applies to table size, so the behavior I describe should be considered as the expected one.

In other words: __what is exactly the issue with Acero? It cannot join _heavy tables_, or it cannot join tables with _heavy keys_?__

**Suggestion of solution**

If the behavior I describe is an actual bug, a potential solution could look like this: 

- Introducing an additional boolean argument in `RowArrayMerge::PrepareForMerge()`. If this argument is TRUE, then the logical test ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L476-L480)) would be performed. If FALSE, it would not be performed;
- This additional argument would be set to TRUE in the first call to `RowArrayMerge::PrepareForMerge()` ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L1333-L1334));
- This additional argument would be set to FALSE in the second call to `RowArrayMerge::PrepareForMerge()` ([here](https://github.com/apache/arrow/blob/e56726c161df1d588f2f6197dbdd0c0011f14855/cpp/src/arrow/acero/swiss_join.cc#L1355-L1357)).

With the help of Chat-GPT, I opened a PR suggesting an (untested) solution ([here](https://github.com/apache/arrow/pull/37709)).

### Component(s)

C++

### PR DESCRIPTION
**Pull request by oliviermeslin**
[PR 35087](https://github.com/apache/arrow/pull/35087) introduced an explicit fail in large joins with Acero when key data is larger than 4GB (solving the problem reported by [issue 34474](https://github.com/apache/arrow/issues/34474)). However, I think (but I'm not sure) that this quick fix is too restrictive because the total size condition is applied to the total size of tables to be joined, rather than to the size of keys. As a consequence, Acero fails when trying to merge large tables, even when the size of key data is well below 4 GB.

This PR modifies the source code so that the logical test only verifies whether the total size of _key variable_ is below 4 GB.

### Rationale for this change

In the current situation, joins with arrow fail when the tables to be joined are jointly larger than 4GB, even if the key data is smaller than 4GB. This is due to the fact that the test on the size of key data is erroneously applied to the size the tables to be joined.

### What changes are included in this PR?

I modify slightly the C++ code so that the test on the size of key data does not apply any more to the size the tables to be joined.

### Are these changes tested?

Not so far.

### Are there any user-facing changes?

No. 

* Closes: #37655

### CODE REVIEW

File: cpp/src/arrow/acero/swiss_join.cc [hunk 1]
```diff
@@ -1331,7 +1332,7 @@ Status SwissTableForJoinBuild::PreparePrtnMerge() {
     partition_keys[i] = prtn_states_[i].keys.keys();
   }
   RETURN_NOT_OK(RowArrayMerge::PrepareForMerge(target_->map_.keys(), partition_keys,
-                                               &partition_keys_first_row_id_, pool_));
+                                               &partition_keys_first_row_id_, pool_, true));
```
Comment (oliviermeslin) [inline cpp/src/arrow/acero/swiss_join.cc:1335]:
```suggestion
                                               &partition_keys_first_row_id_, pool_, 
                                               true));
```

File: cpp/src/arrow/acero/swiss_join_internal.h [hunk 2]
```diff
@@ -161,7 +161,7 @@ class RowArrayMerge {
   //
   static Status PrepareForMerge(RowArray* target, const std::vector<RowArray*>& sources,
                                 std::vector<int64_t>* first_target_row_id,
-                                MemoryPool* pool);
+                                MemoryPool* pool, bool check_key_size);
```
Comment (pitrou) [inline cpp/src/arrow/acero/swiss_join_internal.h:164]: Nits:
1. let's keep the `MemoryPool*` argument at the end
2. perhaps change the argument name to not expose an implementation detail?

So for example:
```c++
  static Status PrepareForMerge(RowArray* target, const std::vector<RowArray*>& sources,
                                bool is_key_data, std::vector<int64_t>* first_target_row_id,
                                MemoryPool* pool);
```

**Comment by pitrou** [comment]
@oliviermeslin Thanks for this. This probably can't be tested efficiently in the test suite unfortunately, due to the test size required.

**Review by raulcd — COMMENTED** [review]
There seems to be some lint failures: https://github.com/apache/arrow/actions/runs/6470238300/job/17569780906?pr=37709#step:5:864

File: cpp/src/arrow/acero/swiss_join.cc [hunk 3]
```diff
@@ -1331,7 +1332,8 @@ Status SwissTableForJoinBuild::PreparePrtnMerge() {
     partition_keys[i] = prtn_states_[i].keys.keys();
   }
   RETURN_NOT_OK(RowArrayMerge::PrepareForMerge(target_->map_.keys(), partition_keys,
-                                               &partition_keys_first_row_id_, pool_));
+                                               &partition_keys_first_row_id_, pool_, 
+                                               true));
```
Comment (oliviermeslin) [inline cpp/src/arrow/acero/swiss_join.cc:1336]:
```suggestion
                                               &partition_keys_first_row_id_, pool_,
                                               true));
```

**Comment by oliviermeslin** [comment]
@raulcd : I think I solved the problem (I forgot to remove a trailing whitespace). Can you please re-launch the checks?

File: cpp/src/arrow/acero/swiss_join_internal.h (hunk 2 above)
Comment (oliviermeslin) [reply cpp/src/arrow/acero/swiss_join_internal.h]: Thanks for your review! I think I adjusted my PR accordingly.

**Comment by westonpace** [comment]
@oliviermeslin I agree it's too big to test in any kind of reasonable unit test.  have you run any tests manually with payloads larger than 4GB and confirmed the results are correct?

**Review by westonpace — COMMENTED** [review]
I can't verify that joins with 4GiB or more of payload data will work correctly, I never tested for that.  But I agree that the issue I fixed was specific to key and so, assuming we have manually done some testing of large payload data then it seems fine to loosen this restriction.

**Comment by raulcd** [comment]
> @raulcd : I think I solved the problem (I forgot to remove a trailing whitespace). Can you please re-launch the checks?

I've pushed a minor lint change, I do think that this should fix it. GitHub UI did not allowed me to propose the change to remove a trailing whitespace, that's why I just pushed myself.

**Review by westonpace — APPROVED** [review]
The only failing test now appears to be unrelated (s3).  Conditional approval provided someone has at least done some basic manual testing to verify that this does actually work.

**Comment by oliviermeslin** [comment]
Thanks @westonpace ! I already prepared some tests (one using artificial data, the other one using real data). Unfortunately, I could not install `arrow` as an `R` _package_ from my own fork, because installing a development version of `arrow` seems to be quite complicated (see [this issue](https://github.com/apache/arrow/issues/35472)). I'll try my best to find a solution and I'll let you know the result.

**Comment by amoeba** [comment]
Hey @oliviermeslin, I ran your example from https://github.com/apache/arrow/issues/37655 w/o this patch and got the expected error after the script printed "Doing the join with 9 variables":

> ! Invalid: There are more than 2^32 bytes of key data.  Acero cannot process a join of this magnitude

I then built libarrow and the R package off your patch and actually got a segfault. Full output below:

<details>
<summary>Output with segfault</summary>

```
❯ Rscript acero-join-test.R
Welcome to R :)
This session's PID is 2890
Some features are not enabled in this build of Arrow. Run `arrow_info()` for more information.

Attaching package: ‘arrow’

The following object is masked from ‘package:utils’:

    timestamp


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

[1] "Doing the join with 2 variables"
[1] "id"        "variable1"
[1] "Doing the join with 3 variables"
[1] "id"        "variable1" "variable2"
[1] "Doing the join with 4 variables"
[1] "id"        "variable1" "variable2" "variable3"
[1] "Doing the join with 5 variables"
[1] "id"        "variable1" "variable2" "variable3" "variable4"
[1] "Doing the join with 6 variables"
[1] "id"        "variable1" "variable2" "variable3" "variable4" "variable5"
[1] "Doing the join with 7 variables"
[1] "id"        "variable1" "variable2" "variable3" "variable4" "variable5"
[7] "variable6"
[1] "Doing the join with 8 variables"
[1] "id"        "variable1" "variable2" "variable3" "variable4" "variable5"
[7] "variable6" "variable7"
[1] "Doing the join with 9 variables"
[1] "id"        "variable1" "variable2" "variable3" "variable4" "variable5"
[7] "variable6" "variable7" "variable8"

 *** caught segfault ***
address 0x7ed0457b20, cause 'invalid permissions'

 *** caught segfault ***
address 0x7f6676b420, cause 'invalid permissions'

 *** caught segfault ***
address 0x7ed8992de0, cause 'invalid permissions'

 *** caught segfault ***
address 0x7ee948e360, cause 'invalid permissions'

Traceback:
 1: Table__from_ExecPlanReader(self)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader)

Traceback:
 1: Table__from_ExecPlanReader(self)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader)
 4: as_arrow_table(reader)
 5: as_arrow_table.arrow_dplyr_query(x)
 6: as_arrow_table(x)

Traceback:

Traceback:
 1: Table__from_ExecPlanReader(self) 4: as_arrow_table(reader)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader) 7: doTryCatch(return(expr), name, parentenv, handler)
 8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
 5:  1:
as_arrow_table.arrow_dplyr_query(x)
 6: as_arrow_table(x) 4:
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: as_arrow_table(reader)
 5: as_arrow_table.arrow_dplyr_query(x)Table__from_ExecPlanReader(self)
 2: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, data(all_of(vars_temp),     all_of(vars_temp)), by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...

10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())(x$read_table())()
 6: as_arrow_table(x)
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: })
tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)

 3: as_arrow_table.RecordBatchReader(reader)
 4: as_arrow_table(reader)10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {
 5: as_arrow_table.arrow_dplyr_query(x)
 6:     augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, select(data, all_of(vars_temp)),     by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...
as_arrow_table(x)
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, select(data, all_of(vars_temp)),     by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, "Doing thes"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...
11: compute.arrow_dplyr_query(left_join(data, vars_temp(), by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...

 *** caught segfault ***
address 0x7f6ecc2520, cause 'invalid permissions'

Traceback:
 1: Table__from_ExecPlanReader(self)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader)
 4: as_arrow_table(reader)
 5: as_arrow_table.arrow_dplyr_query(x)
 6: as_arrow_table(x)
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, select(data, all_of(vars_temp)),     by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...

 *** caught segfault ***
address 0x7f5e1daf00, cause 'invalid permissions'

Traceback:
 1: Table__from_ExecPlanReader(self)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader)
 4: as_arrow_table(reader)
 5: as_arrow_table.arrow_dplyr_query(x)
 6: as_arrow_table(x)
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, select(data, all_of(vars_temp)),     by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...

 *** caught segfault ***
address 0x7ee0f0bac0, cause 'invalid permissions'

Traceback:
 1: Table__from_ExecPlanReader(self)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader)
 4: as_arrow_table(reader)
 5: as_arrow_table.arrow_dplyr_query(x)
 6: as_arrow_table(x)
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, select(data, all_of(vars_temp)),     by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...

 *** caught segfault ***
address 0x7ecb394000, cause 'invalid permissions'

Traceback:
 1: Table__from_ExecPlanReader(self)
 2: x$read_table()
 3: as_arrow_table.RecordBatchReader(reader)
 4: as_arrow_table(reader)
 5: as_arrow_table.arrow_dplyr_query(x)
 6: as_arrow_table(x)
 7: doTryCatch(return(expr), name, parentenv, handler)
 8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
 9: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(as_arrow_table(x), error = function(e, call = caller_env(n = 4)) {    augment_io_error_msg(e, call, schema = schema())})
11: compute.arrow_dplyr_query(left_join(data, select(data, all_of(vars_temp)),     by = c(id = "id")))
12: compute(left_join(data, select(data, all_of(vars_temp)), by = c(id = "id")))
13: FUN(X[[i]], ...)
14: lapply(1:nb_var, function(n) {    print(paste0("Doing the join with ", n + 1, " variables"))    vars_temp <- c("id", vars[1:n])    print(vars_temp)    data_out <- compute(left_join(data, select(data, all_of(vars_temp)),         by = c(id = "id")))    return("Success!")})
An irrecoverable exception occurred. R is aborting now ...
fish: Job 1, 'Rscript acero-join-test.R' terminated by signal SIGSEGV (Address boundary error)
```
</details>

**Comment by oliviermeslin** [comment]
@amoeba : I could finally install arrow with my patch (on Linux/Ubuntu). I re-ran my my patch with using the `gdb` debugger and got the same error as you. The debugger output is the following:

```
arrow::acero::RowArrayMerge::CopyVaryingLength (target=<optimized out>, source=..., first_target_row_id=<optimized out>, first_target_row_offset=<optimized out>, source_rows_permutation=0x7ffcc8052020) at /home/onyxia/work/arrow/cpp/src/arrow/acero/swiss_join.cc:605
605             *target_row_ptr++ = *source_row_ptr++;
```

[Line 605 of swiss_join.cc](https://github.com/oliviermeslin/arrow/blob/fb2617833200d9c90b6d79a85ae9e840b41938da/cpp/src/arrow/acero/swiss_join.cc#L604) is part of the following loop, where the content of source rows is iteratively added to target rows:

```
for (uint32_t word = 0; word < length / sizeof(uint64_t); ++word) {
  *target_row_ptr++ = *source_row_ptr++;
}
```

As I am completly new to C++, I'm not sure what this bug means.  My intuition is that we're trying to add too much to the `target_row_ptr` pointer given its type. Would you or @westonpace have any clue on what is going wrong?

File: cpp/src/arrow/acero/swiss_join.cc [hunk 4]
```diff
@@ -473,7 +474,7 @@ Status RowArrayMerge::PrepareForMerge(RowArray* target,
     (*first_target_row_id)[sources.size()] = num_rows;
   }
 
-  if (num_bytes > std::numeric_limits<uint32_t>::max()) {
+  if (is_key_data && num_bytes > std::numeric_limits<uint32_t>::max()) {
```
Comment (zanmato1984) [inline cpp/src/arrow/acero/swiss_join.cc:477]: When letting non-key data that is greater than `std::numeric_limits<uint32_t>::max()` bypass this check, the `num_bytes` will underflow to a much smaller value in the `static_cast<uint32_t>` in #486. Then the `target` won't allocate enough space, resulting in segfault when copying data to target.

**Review by zanmato1984 — CHANGES_REQUESTED** [review]

**Comment by zanmato1984** [comment]
Hi @oliviermeslin @amoeba , I was trying to help on the issue you met. After looking a bit, I'd say it's unfortunately an inherent problem of the fix. In other words, the original issue just transfers to a later place and explodes in a more implicit fashion (the segfault). 

Please see my comment on the changed code about what is actually happening. Thanks.

**Comment by zanmato1984** [comment]
PR #43389 fixed the issue in a more thorough way so I'm closing this one. Thanks.

**Pull request closed without merging**